cpp/src/explainer/tree_shap.cu

```diff
@@ -0,0 +1,326 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuml/explainer/tree_shap.hpp>
+#include <raft/error.hpp>
+#include <thrust/device_ptr.h>
+#include <GPUTreeShap/gpu_treeshap.h>
+#include <treelite/tree.h>
+#include <memory>
+#include <type_traits>
+#include <iostream>
+#include <vector>
+
+namespace {
+
+namespace tl = treelite;
+
+// Define a custom split condition implementing EvaluateSplit and Merge
+template <typename T>
+struct MySplitCondition {
+  MySplitCondition() = default;
+  MySplitCondition(T feature_lower_bound, T feature_upper_bound)
+      : feature_lower_bound(feature_lower_bound),
+        feature_upper_bound(feature_upper_bound) {
+    assert(feature_lower_bound <= feature_upper_bound);
+  }
+
+  /*! Feature values >= lower and < upper flow down this path. */
+  T feature_lower_bound;
+  T feature_upper_bound;
+
+  // Does this instance flow down this path?
+  __host__ __device__ bool EvaluateSplit(T x) const {
+    return x >= feature_lower_bound && x < feature_upper_bound;
+  }
+
+  // Combine two split conditions on the same feature
+  __host__ __device__ void Merge(
+      const MySplitCondition& other) {  // Combine duplicate features
+    feature_lower_bound = max(feature_lower_bound, other.feature_lower_bound);
+    feature_upper_bound = min(feature_upper_bound, other.feature_upper_bound);
+  }
+  static_assert(std::is_same<T, float>::value || std::is_same<T, double>::value,
+                "T must be a float or double");
+};
+
+enum class CondType : uint8_t {
+  kFloat, kDouble
+};
+
+class ExtractedPath {
+ public:
+  CondType cond_type;
+  int num_tree;
+  float global_bias;
+  tl::TaskType task_type;
+  tl::TaskParam task_param;
+  bool average_tree_output;
+
+  virtual CondType GetCondType() = 0;
+  virtual ~ExtractedPath() = default;
+
+  template <typename T>
+  static std::unique_ptr<ExtractedPath> Create();
+
+  template <typename Func, typename ...Args>
+  auto Dispatch(Func func, Args&& ...args);
+};
+
+template <typename T>
+class ExtractedPathImpl : public ExtractedPath {
+ public:
+  std::vector<gpu_treeshap::PathElement<MySplitCondition<T>>> paths;
+  ExtractedPathImpl() {
+    if (std::is_same<T, double>::value) {
+      cond_type = CondType::kDouble;
+    } else {
+      cond_type = CondType::kFloat;
+    }
+  }
+  virtual ~ExtractedPathImpl() = default;
+
+  CondType GetCondType() override {
+    return cond_type;
+  }
+};
+
+template <typename T>
+std::unique_ptr<ExtractedPath> ExtractedPath::Create() {
+  std::unique_ptr<ExtractedPath> model
+    = std::make_unique<ExtractedPathImpl<T>>();
+  return model;
+}
+
+template <typename Func, typename ...Args>
+auto
+ExtractedPath::Dispatch(Func func, Args&& ...args) {
```

Comment:
I'm not immediately seeing why the dispatch and additional indirection of `ExtractedPath`/`ExtractedPathImpl` is necessary. Could you add a comment in-code to describe what prompts this?

Reply:
Got rid of the dispatch function entirely.